google/cloud/spanner/interval.cc

```diff
@@ -0,0 +1,367 @@
+  }
+
+  static absl::string_view ConsumeSign(absl::string_view& s) {
+    auto t = s;
+    if (!absl::ConsumePrefix(&s, "-")) absl::ConsumePrefix(&s, "+");
+    return t.substr(0, t.size() - s.size());
+  }
+
+  static absl::string_view ConsumeWord(absl::string_view& s) {
+    auto t = s;
+    while (!s.empty() && absl::ascii_isalpha(s.front())) s.remove_prefix(1);
+    return t.substr(0, t.size() - s.size());
+  }
+
+  static bool ParseDouble(absl::string_view& s, double& d, bool allow_sign) {
+    auto t = s;
+    if (allow_sign) ConsumeSign(t);
+    ConsumeInteger(t);
+    if (!absl::ConsumePrefix(&t, ".")) return false;
+    ConsumeInteger(t);
+    auto len = s.size() - t.size();
+    if (!absl::SimpleAtod(s.substr(0, len), &d)) return false;
+    s.remove_prefix(len);
+    return true;
+  }
+
+  template <typename T>
+  static bool ParseInteger(absl::string_view& s, T& n, bool allow_sign) {
+    auto t = s;
+    if (allow_sign) ConsumeSign(t);
+    ConsumeInteger(t);
+    auto len = s.size() - t.size();
+    if (!absl::SimpleAtoi(s.substr(0, len), &n)) return false;
+    s.remove_prefix(len);
+    return true;
+  }
+
+  static bool ParseYearMonth(absl::string_view& s, Interval& intvl) {
+    auto t = s;
+    std::int32_t year;
+    if (!ParseInteger(t, year, true)) return false;
+    if (!absl::ConsumePrefix(&t, "-")) return false;
+    std::int32_t month;
+    if (!ParseInteger(t, month, false)) return false;
+    intvl = Interval(year, month, 0);
+    s.remove_prefix(s.size() - t.size());
+    return true;
+  }
+
+  static bool ParseHourMinuteSecond(absl::string_view& s, Interval& intvl) {
+    auto t = s;
+    auto d = absl::ZeroDuration();
+    auto sign = ConsumeSign(t);
+    std::int64_t hour;
+    if (!ParseInteger(t, hour, false)) return false;
+    d += absl::Hours(hour);
+    if (!absl::ConsumePrefix(&t, ":")) return false;
+    std::int64_t minute;
+    if (!ParseInteger(t, minute, false)) return false;
+    d += absl::Minutes(minute);
+    if (!absl::ConsumePrefix(&t, ":")) return false;
+    double second;
+    if (ParseDouble(t, second, false)) {
+      d += absl::Seconds(second);
+    } else {
+      std::int64_t second;
+      if (!ParseInteger(t, second, false)) return false;
+      d += absl::Seconds(second);
+    }
+    intvl = Interval(sign == "-" ? -d : d);
+    s.remove_prefix(s.size() - t.size());
+    return true;
+  }
+
+  static Interval Unit(absl::string_view& s, std::int32_t n) {
+    auto t = s;
+    auto unit = ConsumeWord(t);
+    if (unit.size() > 1) absl::ConsumeSuffix(&unit, "s");  // unpluralize
+    for (auto const& factory : kUnitFactories) {
+      if (absl::StartsWith(factory.name, unit)) {
+        s.remove_prefix(s.size() - t.size());
+        return factory.factory(n);
+      }
+    }
+    return Interval(0, 0, n);  // default to days without removing unit
+  }
+
+  std::string s_;
+};
+
+StatusOr<Interval> Parser::Parse() {
+  Interval intvl;
+  auto s = absl::string_view(s_);
```

Comment:
Do you even need the `Parser` class?  This is the only place you use `s_`, and it seems you could just say:

```cc
StatusOr<Interval> ParseInterval(std::string s_) { 
  Interval intvl;
  auto s = absl::string_view(s_);
 ...
```